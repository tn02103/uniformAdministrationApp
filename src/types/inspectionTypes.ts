import { Prisma } from "@prisma/client";
import { cadetLableArgs } from "./globalCadetTypes";

export const plannedInspectionTypeArgs = Prisma.validator<Prisma.InspectionFindManyArgs>()({
    include: {
        deregistrations: {
            include: {
                cadet: cadetLableArgs
            },
        },
    }
});
export type PlannedInspectionType = Prisma.InspectionGetPayload<typeof plannedInspectionTypeArgs>;
