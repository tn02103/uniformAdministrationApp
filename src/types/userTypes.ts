import { Prisma } from "@prisma/client";

export const userArgs = Prisma.validator<Prisma.UserFindManyArgs>()({
    select: {
        id: true,
        username: true,
        name: true,
        active: true,
        role: true
    },
    orderBy: { username: "asc"}
});

export type User = Prisma.UserGetPayload<typeof userArgs>
