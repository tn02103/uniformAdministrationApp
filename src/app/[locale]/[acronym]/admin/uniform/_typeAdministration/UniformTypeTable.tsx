"use client"

import { TooltipActionButton } from "@/components/Buttons/TooltipIconButton";
import { ReorderableTableBody } from "@/components/reorderDnD/ReorderableTableBody";
import { changeUniformTypeSortOrder } from "@/dal/uniform/type/_index";
import { useUniformTypeList } from "@/dataFetcher/uniformAdmin";
import { useI18n } from "@/lib/locales/client";
import { UniformType } from "@/types/globalUniformTypes";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useState } from "react";
import { Table } from "react-bootstrap";
import { toast } from "react-toastify";
import { UniformTypeOffcanvas } from "./UniformTypeOffcanvas";

const listElementStyles = {
    padding: "10px",
    margin: "0px",
}
export const UniformTypeTable = (props: { initialTypeList: UniformType[] }) => {
    const t = useI18n();

    const { typeList, mutate } = useUniformTypeList(props.initialTypeList);
    const [selectedTypeId, setSelectedTypeId] = useState<string | null>(null);
    const [editable, setEditable] = useState(false);

    const handleChangeSortorder = async (newArray: UniformType[], itemId: string) => {
        if (newArray.length !== typeList?.length) return;

        const newPosition = newArray.findIndex(i => i.id === itemId);
        if (newPosition === -1) return;

        await mutate(
            changeUniformTypeSortOrder({ typeId: itemId, newPosition }),
            {
                optimisticData: newArray
            }
        ).then(() => {
            toast.success(t('common.success.changeSortorder'));
        }).catch((e) => {
            console.error(e);
            toast.error(t('common.error.actions.changeSortorder'));
        });
    }

    return (
        <div>
            <Table striped aria-label={t('common.uniform.type.type', { count: 2 })}>
                <thead>
                    <tr>
                        <th></th>
                        <th>{t('common.uniform.type.type', { count: 1 })}</th>
                        <th>{t('common.uniform.type.acronym')}</th>
                        <th className="d-none d-md-table-cell">{t('common.uniform.type.issuedDefault')}</th>
                        <th className="d-none d-sm-table-cell">{t('common.uniform.type.usingSizes')}</th>
                        <th className="d-none d-sm-table-cell">{t('common.uniform.type.usingGenerations')}</th>
                        <th className="d-none d-lg-table-cell">{t('common.uniform.type.defaultSizelist')}</th>
                        <th className="d-none d-lg-table-cell">{t('common.uniform.generation.label', { count: 2 })}</th>
                        <th>
                            <TooltipActionButton
                                variantKey="create"
                                disabled={editable}
                                onClick={() => { setSelectedTypeId('new'); setEditable(true); }} />
                        </th>
                    </tr>
                </thead>
                <ReorderableTableBody<UniformType>
                    items={typeList ?? []}
                    itemType="UNIFORM_TYPE"
                    onDragEnd={handleChangeSortorder}
                >
                    {({ draggableRef, previewRef, isDragging, item }) => (
                        <tr key={item.id} ref={previewRef} aria-label={item.name} style={isDragging ? { ...listElementStyles, opacity: 0 } : listElementStyles}>
                            <td>
                                <span ref={draggableRef} aria-label="move Item" className="p-2">
                                    <FontAwesomeIcon
                                        icon={faBars}
                                        className="text-seccondary"
                                    />
                                </span>
                            </td>
                            <td>
                                {item.name}
                            </td>
                            <td>{item.acronym}</td>
                            <td className="d-none d-md-table-cell">{item.issuedDefault}</td>
                            <td className="d-none d-sm-table-cell">{item.usingSizes ? t('common.yes') : t('common.no')}</td>
                            <td className="d-none d-sm-table-cell">{item.usingGenerations ? t('common.yes') : t('common.no')}</td>
                            <td className="d-none d-lg-table-cell">{item.defaultSizelist?.name}</td>
                            <td className="d-none d-lg-table-cell">{item.uniformGenerationList.length}</td>
                            <td>
                                <TooltipActionButton
                                    variantKey="open"
                                    disabled={editable}
                                    onClick={() => setSelectedTypeId(item.id)} />
                            </td>
                        </tr>
                    )}
                </ReorderableTableBody>
            </Table>
            {
                selectedTypeId && (
                    <UniformTypeOffcanvas
                        editable={editable}
                        setEditable={setEditable}
                        uniformType={typeList?.find(t => t.id === selectedTypeId) ?? null}
                        setSelectedTypeId={setSelectedTypeId}
                    />
                )
            }
        </div>
    );
}
