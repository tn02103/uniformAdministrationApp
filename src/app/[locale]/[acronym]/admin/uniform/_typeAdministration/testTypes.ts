import { UniformType } from "@/types/globalUniformTypes";

const sizeListIds = [
    'e667d674-7df8-436b-a2b8-77b06e063d36',
    'a961545b-28a7-409e-9200-1d85ccd53522',
    '07de1d59-4fc6-447b-98a6-da916e5792ef',
];
export const testTypes: UniformType[] = [
    {
        id: "fbe495ec-799e-46c9-8aa5-ca34e1447bf4",
        name: "Test Type",
        acronym: "TT",
        sortOrder: 1,
        usingSizes: true,
        usingGenerations: true,
        issuedDefault: 2,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            },
            {
                id: "7feb435c-ee23-486d-af9f-b7c874383e22",
                name: "Test Generation 2",
                outdated: false,
                sortOrder: 2,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            },
            {
                id: "c46adb18-1a04-4dda-884d-972a7aa0e0da",
                name: "Test Generation 3",
                outdated: true,
                sortOrder: 3,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            },
        ],
    }, {
        id: "c0f5f7d4-5e8b-4c1b-9f2a-7d7e7c6e3f1d",
        name: "Test Type 2",
        acronym: "TT2",
        sortOrder: 2,
        usingSizes: false,
        usingGenerations: false,
        issuedDefault: 1,
        fk_defaultSizelist: sizeListIds[2],
        defaultSizelist: {
            id: sizeListIds[2],
            name: "Test Size List 3",
        },
        uniformGenerationList: [],
    }, {
        id: "26c469f7-aae4-4398-8aba-5d131178c6f8",
        name: "Test Type 3",
        acronym: "TT3",
        sortOrder: 3,
        usingSizes: true,
        usingGenerations: false,
        issuedDefault: 0,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List 1",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            }
        ],
    }, {
        id: "d1e8b3c2-5a2b-4f8b-9c6e-7d7e7c6e3f1d",
        name: "Test Type 4",
        acronym: "TT4",
        sortOrder: 4,
        usingSizes: true,
        usingGenerations: true,
        issuedDefault: 0,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List 1",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            },
            {
                id: "7feb435c-ee23-486d-af9f-b7c874383e22",
                name: "Test Generation 2",
                outdated: false,
                sortOrder: 2,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            }
        ],
    }
]