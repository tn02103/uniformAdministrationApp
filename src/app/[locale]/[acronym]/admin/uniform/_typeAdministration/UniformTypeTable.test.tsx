import { UniformType } from "@/types/globalUniformTypes";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { testTypes } from "./testTypes";
import { UniformTypeTable } from "./UniformTypeTable";


// ################## MOCKS ##################
jest.mock("@/dal/uniform/type/_index", () => ({
    changeUniformTypeSortOrder: jest.fn(async () => "uniform type sortOrder changed"),
}));

jest.mock("@/dataFetcher/uniformAdmin", () => {
    const mutateMock = jest.fn(async (a,) => await a);
    return {
        useUniformTypeList: jest.fn(() => ({
            mutate: mutateMock,
            typeList: testTypes,
        })),
    }
});
/* eslint-disable @typescript-eslint/no-explicit-any */
let onDragEndFunction: undefined | ((newArray: UniformType[], itemId: string) => Promise<any>) = undefined;
jest.mock("@/components/reorderDnD/ReorderableTableBody", () => {
    return {
        ReorderableTableBody: jest.fn(({ items, onDragEnd, children }) => {
            onDragEndFunction = onDragEnd;
            return (
                <tbody data-testid="reorderable-table-body">
                    {items.map(((item: any) => children({ item, draggableRef: undefined, previewRef: undefined, isDragging: false })))}
                </tbody>
            )
        }),
    };
});
/* eslint-enable @typescript-eslint/no-explicit-any */

jest.mock("./UniformTypeOffcanvas", () => ({
    UniformTypeOffcanvas: jest.fn(({ setSelectedTypeId }) => (
        <div data-testid="uniform-type-offcanvas" onClick={() => setSelectedTypeId(null)}>
            UniformTypeOffcanvas
        </div>
    )),
}));


// ################## TESTS ##################
describe("<UniformTypeTable />", () => {
    const { changeUniformTypeSortOrder } = jest.requireMock("@/dal/uniform/type/_index");
    const { useUniformTypeList } = jest.requireMock("@/dataFetcher/uniformAdmin");
    const { UniformTypeOffcanvas } = jest.requireMock("./UniformTypeOffcanvas");
    const { toast } = jest.requireMock("react-toastify");
    const mockMutate = useUniformTypeList().mutate;

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders the component", () => {
        render(<UniformTypeTable initialTypeList={testTypes} />);

        expect(screen.getByRole("table")).toBeInTheDocument();
        expect(screen.getByRole("table").closest('div')).toMatchSnapshot();
    });

    it("calls changeUniformTypeSortOrder when onDragEnd is triggered", async () => {
        changeUniformTypeSortOrder.mockReturnValue("uniform type sortOrder changed");
        render(<UniformTypeTable initialTypeList={testTypes} />);

        expect(onDragEndFunction).toBeDefined();
        const newArray = [
            testTypes[1],
            testTypes[2],
            testTypes[0],
            testTypes[3],
        ]
        await onDragEndFunction?.(newArray, testTypes[0].id);

        expect(changeUniformTypeSortOrder).toHaveBeenCalledWith({ typeId: testTypes[0].id, newPosition: 2 });
        expect(mockMutate).toHaveBeenCalledWith("uniform type sortOrder changed", { optimisticData: newArray });

        expect(toast.success).toHaveBeenCalledTimes(1);
        expect(toast.success).toHaveBeenCalledWith("common.success.changeSortorder");
    });

    it('catchs exeption in changeUniformTypeSortOrder', async () => {
        changeUniformTypeSortOrder.mockImplementation(async () => { throw new Error("error") });
        render(<UniformTypeTable initialTypeList={testTypes} />);

        expect(onDragEndFunction).toBeDefined();
        const newArray = [
            testTypes[1],
            testTypes[2],
            testTypes[0],
            testTypes[3],
        ]
        await onDragEndFunction?.(newArray, testTypes[0].id);


        expect(changeUniformTypeSortOrder).toHaveBeenCalledWith({ typeId: testTypes[0].id, newPosition: 2 });
        expect(mockMutate).not.toHaveBeenCalledWith("uniform type sortOrder changed", { optimisticData: newArray });

        expect(toast.error).toHaveBeenCalledTimes(1);
        expect(toast.error).toHaveBeenCalledWith("common.error.actions.changeSortorder");
    });

    it("does not call sortOrder function when itemId is not in the list", () => {
        render(<UniformTypeTable initialTypeList={testTypes} />);

        expect(onDragEndFunction).toBeDefined();
        onDragEndFunction?.([
            testTypes[1],
            testTypes[2],
            testTypes[0],
            testTypes[3],
        ], "invalid-id");

        expect(changeUniformTypeSortOrder).not.toHaveBeenCalled();
    });
    it("does not call sortOrder function when list has wrong size", () => {
        render(<UniformTypeTable initialTypeList={testTypes} />);

        expect(onDragEndFunction).toBeDefined();
        onDragEndFunction?.([
            testTypes[1],
            testTypes[2],
            testTypes[0],
        ], testTypes[0].id);

        expect(changeUniformTypeSortOrder).not.toHaveBeenCalled();
    });

    describe("Render UniformTypeOffcanvas", () => {
        it("opens and closes UniformTypeOffcanvas for the correct element", async () => {
            const user = userEvent.setup();
            render(<UniformTypeTable initialTypeList={testTypes} />);

            // open the offcanvas to edit the second type
            const openButtons = screen.getAllByRole('button', { name: 'open' });
            expect(openButtons).toHaveLength(4);
            await user.click(openButtons[1]);

            // check if the offcanvas is opened with the correct type
            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(1);
            expect(UniformTypeOffcanvas).toHaveBeenCalledWith(
                {
                    editable: false,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[1]
                },
                undefined
            );

            // close the offcanvas by calling the setSelectedTypeId function with null
            act(() => {
                UniformTypeOffcanvas.mock.calls[0][0].setSelectedTypeId(null);
            });
            expect(screen.queryByTestId("uniform-type-offcanvas")).not.toBeInTheDocument();

            // open the offcanvas for the forth type
            await user.click(openButtons[3]);
            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(2);
            expect(UniformTypeOffcanvas).toHaveBeenLastCalledWith(
                {
                    editable: false,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[3]
                },
                undefined
            );
        });

        it("opens UniformTypeOffcanvas to create a new type", async () => {
            const user = userEvent.setup();
            render(<UniformTypeTable initialTypeList={testTypes} />);

            const createButton = screen.getByRole('button', { name: 'create' });
            await user.click(createButton);

            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(1);
            expect(UniformTypeOffcanvas).toHaveBeenCalledWith(
                {
                    editable: true,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: null
                },
                undefined
            );

            act(() => {
                UniformTypeOffcanvas.mock.calls[0][0].setSelectedTypeId(null);
            });
            expect(screen.queryByTestId("uniform-type-offcanvas")).not.toBeInTheDocument();
        });

        it("rerenders when selectedTypeId changes", async () => {
            const user = userEvent.setup();
            render(<UniformTypeTable initialTypeList={testTypes} />);

            const createButton = screen.getByRole('button', { name: 'create' });
            await user.click(createButton);

            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(1);

            act(() => {
                UniformTypeOffcanvas.mock.calls[0][0].setSelectedTypeId(testTypes[1].id);
            });
            expect(screen.queryByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(2);
            expect(UniformTypeOffcanvas).toHaveBeenCalledWith(
                {
                    editable: true,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[1]
                },
                undefined
            );

            act(() => {
                UniformTypeOffcanvas.mock.calls[1][0].setSelectedTypeId(null);
            });
            expect(screen.queryByTestId("uniform-type-offcanvas")).not.toBeInTheDocument();
        });

        it('should change editable when calling function', async () => {
            const user = userEvent.setup();
            render(<UniformTypeTable initialTypeList={testTypes} />);

            // open the offcanvas to edit the second type
            const openButtons = screen.getAllByRole('button', { name: 'open' });
            expect(openButtons).toHaveLength(4);

            await user.click(openButtons[1]);
            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(1);
            expect(UniformTypeOffcanvas).toHaveBeenCalledWith(
                {
                    editable: false,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[1]
                },
                undefined
            );

            // call the setEditable function to change the editable state
            act(() => {
                UniformTypeOffcanvas.mock.calls[0][0].setEditable(true);
            });
            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(2);
            expect(UniformTypeOffcanvas).toHaveBeenLastCalledWith(
                {
                    editable: true,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[1]
                },
                undefined
            );

            // validate open and create buttons are disabled
            const openButtonsOnEdit = screen.getAllByRole('button', { name: 'open' });
            const createButtonOnEdit = screen.getByRole('button', { name: 'create' });
            openButtonsOnEdit.forEach(button => {
                expect(button).toBeDisabled();
            });
            expect(createButtonOnEdit).toBeDisabled();

            // call the setEditable function to change the editable state back to false
            act(() => {
                UniformTypeOffcanvas.mock.calls[1][0].setEditable(false);
            });
            expect(screen.getByTestId("uniform-type-offcanvas")).toBeInTheDocument();
            expect(UniformTypeOffcanvas).toHaveBeenCalledTimes(3);
            expect(UniformTypeOffcanvas).toHaveBeenLastCalledWith(
                {
                    editable: false,
                    setEditable: expect.any(Function),
                    setSelectedTypeId: expect.any(Function),
                    uniformType: testTypes[1]
                },
                undefined
            );

            // validate open and create buttons are enabled again
            const openButtonsAfterEdit = screen.getAllByRole('button', { name: 'open' });
            const createButtonAfterEdit = screen.getByRole('button', { name: 'create' });
            openButtonsAfterEdit.forEach(button => {
                expect(button).not.toBeDisabled();
            });
            expect(createButtonAfterEdit).not.toBeDisabled();
        });
    });
});
