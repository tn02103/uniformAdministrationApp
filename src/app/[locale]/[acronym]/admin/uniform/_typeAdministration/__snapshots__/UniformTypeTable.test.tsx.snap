// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UniformTypeTable /> should render the component correctly 1`] = `
<div>
  <table
    aria-label="common.uniform.type.type"
    class="table table-striped"
  >
    <thead>
      <tr>
        <th />
        <th>
          common.uniform.type.type
        </th>
        <th>
          common.uniform.type.acronym
        </th>
        <th
          class="d-none d-md-table-cell"
        >
          common.uniform.type.issuedDefault
        </th>
        <th
          class="d-none d-sm-table-cell"
        >
          common.uniform.type.usingSizes
        </th>
        <th
          class="d-none d-sm-table-cell"
        >
          common.uniform.type.usingGenerations
        </th>
        <th
          class="d-none d-lg-table-cell"
        >
          common.uniform.type.defaultSizelist
        </th>
        <th
          class="d-none d-lg-table-cell"
        >
          common.uniform.generation.label
        </th>
        <th>
          <button
            aria-label="create"
            class="border-0 align-self-center undefined btn btn-outline-success btn-sm"
            data-testid="btn_create"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-plus "
              data-icon="plus"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
                fill="currentColor"
              />
            </svg>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="reorderable-table-body"
    >
      <tr
        aria-label="Test Type"
        style="padding: 10px; margin: 0px;"
      >
        <td>
          <span
            aria-label="move Item"
            class="p-2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-bars text-seccondary"
              data-icon="bars"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                fill="currentColor"
              />
            </svg>
          </span>
        </td>
        <td>
          Test Type
        </td>
        <td>
          TT
        </td>
        <td
          class="d-none d-md-table-cell"
        >
          2
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.yes
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.yes
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          Test Size List
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          3
        </td>
        <td>
          <button
            aria-label="open"
            class="border-0 align-self-center undefined btn btn-outline-secondary btn-sm"
            data-testid="btn_open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-right-from-square "
              data-icon="arrow-up-right-from-square"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3l0 82.7c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 32C35.8 32 0 67.8 0 112L0 432c0 44.2 35.8 80 80 80l320 0c44.2 0 80-35.8 80-80l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 112c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-320c0-8.8 7.2-16 16-16l112 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
      <tr
        aria-label="Test Type 2"
        style="padding: 10px; margin: 0px;"
      >
        <td>
          <span
            aria-label="move Item"
            class="p-2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-bars text-seccondary"
              data-icon="bars"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                fill="currentColor"
              />
            </svg>
          </span>
        </td>
        <td>
          Test Type 2
        </td>
        <td>
          TT2
        </td>
        <td
          class="d-none d-md-table-cell"
        >
          1
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.no
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.no
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          Test Size List 3
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          0
        </td>
        <td>
          <button
            aria-label="open"
            class="border-0 align-self-center undefined btn btn-outline-secondary btn-sm"
            data-testid="btn_open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-right-from-square "
              data-icon="arrow-up-right-from-square"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3l0 82.7c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 32C35.8 32 0 67.8 0 112L0 432c0 44.2 35.8 80 80 80l320 0c44.2 0 80-35.8 80-80l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 112c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-320c0-8.8 7.2-16 16-16l112 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
      <tr
        aria-label="Test Type 3"
        style="padding: 10px; margin: 0px;"
      >
        <td>
          <span
            aria-label="move Item"
            class="p-2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-bars text-seccondary"
              data-icon="bars"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                fill="currentColor"
              />
            </svg>
          </span>
        </td>
        <td>
          Test Type 3
        </td>
        <td>
          TT3
        </td>
        <td
          class="d-none d-md-table-cell"
        >
          0
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.yes
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.no
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          Test Size List 1
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          1
        </td>
        <td>
          <button
            aria-label="open"
            class="border-0 align-self-center undefined btn btn-outline-secondary btn-sm"
            data-testid="btn_open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-right-from-square "
              data-icon="arrow-up-right-from-square"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3l0 82.7c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 32C35.8 32 0 67.8 0 112L0 432c0 44.2 35.8 80 80 80l320 0c44.2 0 80-35.8 80-80l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 112c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-320c0-8.8 7.2-16 16-16l112 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
      <tr
        aria-label="Test Type 4"
        style="padding: 10px; margin: 0px;"
      >
        <td>
          <span
            aria-label="move Item"
            class="p-2"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-bars text-seccondary"
              data-icon="bars"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                fill="currentColor"
              />
            </svg>
          </span>
        </td>
        <td>
          Test Type 4
        </td>
        <td>
          TT4
        </td>
        <td
          class="d-none d-md-table-cell"
        >
          0
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.yes
        </td>
        <td
          class="d-none d-sm-table-cell"
        >
          common.yes
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          Test Size List 1
        </td>
        <td
          class="d-none d-lg-table-cell"
        >
          2
        </td>
        <td>
          <button
            aria-label="open"
            class="border-0 align-self-center undefined btn btn-outline-secondary btn-sm"
            data-testid="btn_open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-up-right-from-square "
              data-icon="arrow-up-right-from-square"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3l0 82.7c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 32C35.8 32 0 67.8 0 112L0 432c0 44.2 35.8 80 80 80l320 0c44.2 0 80-35.8 80-80l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 112c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-320c0-8.8 7.2-16 16-16l112 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
